lib/docker.ts

```diff
@@ -132,69 +174,93 @@ class Docker {
-              result.txt[filename] = extracted.layers[layerName].txt[filename];
-            }
-          }
-          // go over MD5 sums found in this layer
-          for (const filename of Object.keys(extracted.layers[layerName].md5)) {
-            // file was not found in previous layer
-            if (!Reflect.has(result.md5, filename)) {
-              result.md5[filename] = extracted.layers[layerName].md5[filename];
-            }
-          }
-        }
-      }
-    }
-    return result;
-  }
-
-  /**
-   * Saves the docker image as a TAR file to a temporary location and extract
-   * the specified files from it.
-   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
-   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
-   * @return list of plain text files and list of MD5 sums
+   * Convenience function to wrap save to tar and extract files from tar
+   * @param searchActions array of pattern, callbacks pairs
+   * @returns extracted file products
    */
   public async extract(
-    txtPatterns: string[],
-    md5Patterns: string[] = [],
-  ): Promise<ExtractedKeyFiles> {
+    searchActions: SearchAction[],
+  ): Promise<SearchActionProducts> {
     return this.save(async (err, imageTarPath) => {
       if (err) {
         throw err;
       }
-      return await this.analyze(imageTarPath, txtPatterns, md5Patterns);
+      return await extractFromTar(imageTarPath, searchActions);
     });
   }
+
+  /**
+   * Extract files from image and store their product
+   * @param searchActions array of pattern, callbacks pairs
+   * @returns extracted file products
+   */
+  public async extractAndCache(
+    searchActions: SearchAction[],
+  ): Promise<SearchActionProducts> {
+    try {
+      this.searchActionProducts = Object.assign(
+        this.searchActionProducts,
+        await this.extract(searchActions),
+      );
+      return this.searchActionProducts;
+    } catch (error) {
+      debug(error);
+      return {};
+    }
+  }
+
+  /**
+   * Attempt to perform a static scan
+   * @param searchActions
+   */
+  public async scanStaticalyIfNeeded(searchActions: SearchAction[]) {
+    const size = await this.sizeSafe();
+    if (!size || size > this.GetStaticScanSizeLimit()) {
+      return {};
+    }
+    return await this.extractAndCache(searchActions);
+  }
+
+  /**
+   * Get file product that was previously retrieved and cached or by using
+   *  runtime method after applying the specified callback
+   * @param filename name of file to retrieve its associated string product
+   * @param callbacks optional array of callbacks to call when runtime method
+   *  is used to retrieve the file
+   * @returns map of file products by callback name
+   */
+  public async getFilesProducts(
+    filename: string,
+    searchActionCallback?: SearchActionCallback,
+  ): Promise<{ [key: string]: string | Buffer }> {
```

Comment:
create type for `[searchActionName: string]: string | Buffer` and use it here and in `interface SearchActionProducts` declaration